GeneratorInterface/Core/plugins/ExternalGeneratorFilter.cc

```diff
@@ -287,13 +287,14 @@ void ExternalGeneratorFilter::streamEndRun(edm::StreamID iID, edm::Run const& iR
   }
 }
 void ExternalGeneratorFilter::globalEndRunProduce(edm::Run& iRun, edm::EventSetup const&) const {
-  iRun.emplace(runInfoToken_, std::move(runCache(iRun.index())->runInfo_));
+  iRun.emplace(runInfoToken_, runCache(iRun.index())->runInfo_);
 }
 
 void ExternalGeneratorFilter::globalBeginLuminosityBlockProduce(edm::LuminosityBlock& iLuminosityBlock,
                                                                 edm::EventSetup const&) const {
-  while (not availableForBeginLumi_.load()) {
+  while (state_.load() == State::kNotReadyForNextGlobalBeginLumi) {
   }
+  state_.store(State::kNotReadyForNextGlobalBeginLumi);
```

Comment:
This should probably be a `compare_exchange_strong` just incase there is a race for 2 different Lumis calling `globalBeginLuminosityBlockProduce`. (maybe one Lumi is empty and the next has something and the empty Lumi call was delayed by the OS?)

Reply:
All the ```globalBeginLuminosityBlockProduce``` function calls for a LuminosityBlock (and the other functions in the global begin lumi transition) must complete before the ```EventProcessor``` can move to the next lumi. The ```EventProcessor``` does not even look to see what is next before the entire transition for a lumi is complete. We have to get to the call to ```handleNextEventForStreamAsync``` for another lumi to start. See:

https://cmssdt.cern.ch/dxr/CMSSW/source/FWCore/Framework/src/EventProcessor.cc#1742

I think this is OK.